### PR TITLE
Remote worker and Separate APIs

### PR DESCRIPTION
**Pull request by ortyomka**
### Description of the Change
- Remote worker
- Separate Node and Miner APIs
- More function for miner API

### Benefits
- Better API support
- Other machines can be connected 

### Possible Drawbacks 
- Not track heartbeat of remote worker

### CODE REVIEW

File: core/api/visit.hpp [hunk 1]
```diff
@@ -5,10 +5,67 @@
 
 #pragma once
 
+#include "api/common_api.hpp"
+#include "api/node_api.hpp"
+#include "api/storage_api.hpp"
+#include "api/worker_api.hpp"
+
 namespace fc::api {
```
Comment (turuslan) [inline core/api/visit.hpp:13]:
```suggestion
namespace fc::api {
  struct FullNodeApi;
  struct StorageMinerApi;
  struct WorkerApi;

```

File: core/api/utils.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <future>
+
+#include "markets/storage/ask_protocol.hpp"
```
Comment (turuslan) [inline core/api/utils.hpp:10]:
```suggestion
```

File: core/api/rpc/json.hpp [hunk 3]
```diff
@@ -210,6 +227,30 @@ namespace fc::api {
       decodeEnum(v, j);
     }
 
+    ENCODE(SectorFileType) {
+      return encode(static_cast<int64_t>(v));
```
Comment (turuslan) [inline core/api/rpc/json.hpp:231]:
```suggestion
      return encode(common::to_int(v));
```

File: core/api/rpc/json.hpp [hunk 4]
```diff
@@ -712,6 +761,26 @@ namespace fc::api {
...
+      return j;
+    }
+
+    DECODE(StorageInfo) {
+      decode(v.id, Get(j, "ID"));
+      decode(v.urls, Get(j, "URLs"));
+      decode(v.weight, Get(j, "Weight"));
+      decode(v.can_seal, Get(j, "CanSeal"));
+      decode(v.can_store, Get(j, "CanStore"));
+      decode(v.is_primary, Get(j, "Primary"));
```
Comment (turuslan) [inline core/api/rpc/json.hpp:781]:
```suggestion
      Get(j, "ID", v.id);
      Get(j, "URLs", v.urls);
      Get(j, "Weight", v.weight);
      Get(j, "CanSeal", v.can_seal);
      Get(j, "CanStore", v.can_store);
      Get(j, "Primary", v.is_primary);
```

File: core/api/make.cpp [hunk 5]
```diff
@@ -672,165 +652,160 @@ namespace fc::api {
...
+        }};
+    /**
+     * Payment channel methods are initialized with
+     * PaymentChannelManager::makeApi(Api &api)
+     */
+    api->PaychAllocateLane = {};
+    api->PaychGet = {};
+    api->PaychVoucherAdd = {};
+    api->PaychVoucherCheckValid = {};
+    api->PaychVoucherCreate = {};
```
Comment (turuslan) [inline core/api/make.cpp:808]:
```suggestion
```

File: core/miner/main/main.cpp [hunk 6]
```diff
@@ -303,31 +313,33 @@ namespace fc {
             (path / sector_storage::stores::kMetaFileName).string(),
             *codec::json::format(api::encode(primitives::LocalStorageMeta{
                 uuids::to_string(uuids::random_generator()()),
-                1,
+                10,
```
Comment (Alexey-N-Chernyshov) [inline core/miner/main/main.cpp:316]: Magic number detected, please use named constant.

File: core/remote_worker/main.cpp [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    wsc.setup(*mapi);
+    OUTCOME_TRY(wsc.connect(config.miner_api.first, config.miner_api.second));
+
+    OUTCOME_TRY(version, mapi->Version());
+    // TODO: make version of miner
+    uint64_t miner_api = 0;
+    if (version.api_version != miner_api) {
+      spdlog::error("lotus-miner API version doesn't match: expected: {}",
+                    miner_api);
+      exit(-1);
```
Comment (Alexey-N-Chernyshov) [inline core/remote_worker/main.cpp:143]: EXIT_FAILURE?

File: core/remote_worker/main.cpp [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    auto scheduler{std::make_shared<libp2p::protocol::AsioScheduler>(
+        *io, libp2p::protocol::SchedulerConfig())};
+
+    auto mapi{std::make_shared<api::StorageMinerApi>()};
+    api::rpc::Client wsc{*io};
+    wsc.setup(*mapi);
+    OUTCOME_TRY(wsc.connect(config.miner_api.first, config.miner_api.second));
+
+    OUTCOME_TRY(version, mapi->Version());
+    // TODO: make version of miner
```
Comment (Alexey-N-Chernyshov) [inline core/remote_worker/main.cpp:138]: Is it difficult to add version?

File: core/remote_worker/main.cpp [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        config.join("storage.json"))};
+    OUTCOME_TRY(storage->setStorage([&](auto &config2) {
+      if (config2.storage_paths.empty()) {
+        boost::filesystem::path path{config.join("sectors")};
+        boost::filesystem::create_directories(path);
+        OUTCOME_EXCEPT(common::writeFile(
+            (path / sector_storage::stores::kMetaFileName).string(),
+            *codec::json::format(api::encode(primitives::LocalStorageMeta{
+                uuids::to_string(uuids::random_generator()()),
+                10,
```
Comment (Alexey-N-Chernyshov) [inline core/remote_worker/main.cpp:169]: What is 10?

File: core/remote_worker/main.cpp (hunk 7 above)
Comment (turuslan) [reply core/remote_worker/main.cpp:143]: Also, we may need such function which logs and exits to reuse in other places

File: core/api/visit.hpp (hunk 1 above)
Comment (ortyomka) [reply core/api/visit.hpp]: Fixed

File: core/api/utils.hpp (hunk 2 above)
Comment (ortyomka) [reply core/api/utils.hpp]: Fixed

File: core/api/rpc/json.hpp (hunk 3 above)
Comment (ortyomka) [reply core/api/rpc/json.hpp]: Fixed

File: core/api/rpc/json.hpp (hunk 4 above)
Comment (ortyomka) [reply core/api/rpc/json.hpp]: Fixed

File: core/api/make.cpp (hunk 5 above)
Comment (ortyomka) [reply core/api/make.cpp]: Fixed

File: core/miner/main/main.cpp (hunk 6 above)
Comment (ortyomka) [reply core/miner/main/main.cpp]: Fixed

File: core/remote_worker/main.cpp (hunk 7 above)
Comment (ortyomka) [reply core/remote_worker/main.cpp]: Fixed

File: core/remote_worker/main.cpp (hunk 8 above)
Comment (ortyomka) [reply core/remote_worker/main.cpp:138]: Fixed

File: core/remote_worker/main.cpp (hunk 9 above)
Comment (ortyomka) [reply core/remote_worker/main.cpp]: Fixed

**Review by turuslan — APPROVED** [review]

File: core/remote_worker/main.cpp (hunk 8 above)
Comment (Alexey-N-Chernyshov) [reply core/remote_worker/main.cpp:138]: Does this comment relevant after fix?

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**